rfcs/2020-09-21-0000-kubernetes-metrics-overview.md

```diff
@@ -0,0 +1,283 @@
+  format.
+
+  Among others, this category includes:
+
+  - `kubelet`
+  - `kube-apiserver`
+  - `kube-scheduler`
+  - `kube-controller-manager`
+  - `etcd`
+  - `docker`
+
+  More on this [later](#metrics-exposed-by-the-kubernetes-system-components),
+  but the key part here is those components are not all deployed together and
+  require different access paths if we want to reach their `/metrics` endpoints.
+
+- Metrics available via the [Metrics API] - the `kube-apiserver` API that offers
+  system metrics for the purposes of autoscaling to various Kubernetes
+  components (typically autoscalers -  i.e. Horizontal Pod Autoscaler,
+  Vertical Pod Autoscaler, etc).
+
+  This API essentially contains very limited data on resource usage of the
+  `Pod`s and `Node`s.
+
+  Reviewed in more detail at the Metrics API RFC.
+
+- Metrics gathered from the kernel, at the host level (CPU/RAM/IO usage).
+
+  These metrics themselves are pretty much well-known, they are mostly provided
+  by the kernel.
+
+  However, there are numerous ways to gather them. Most common these days is
+  Prometheus' `node_exporter`, deployed as a dedicated `DaemonSet`.
+  Vector too can gether host metrics - however it might also require a dedicated
+  `DaemonSet` with special capabilities / cgroups (in addition to our log
+  collecting `DaemonSet`) to have proper access to the metrics.
+
+- Metrics derived from the state of the Kubernetes API (i.e. the data at
+  the `kube-apiserver`).
+
+  An implementation of this concept would be the best to explanation:
+  <https://github.com/kubernetes/kube-state-metrics>
+
+  The metrics derived from this state are in the lines of "how many deployments
+  are there" and "how many *healthy* deployments are there".
+
+- Prometheus server providing access to metrics it gathers.
+
+  The presence of a Prometheus server, it being a de-facto standard for metrics
+  aggregation at Kubernetes, it is possible to just grab all the metrics that
+  it exposes. This enables a tool like Vector to abstract away the notion of
+  which metrics in particular are being gathered, but still provide a way to
+  organize the metrics shipping pipeline.
+
+  > Aggregating metrics like Prometheus does it is not trivial, and despite it
+  > being possible to implement Prometheus-like storage and aggregation in
+  > Vector, working with the emerged standard rather than competing with it
+  > would be better for the industry as a whole.
+  > Especially given that Prometheus works well and provides very good
+  > extensibility options, so I don't see technical reasons for competition
+  > here.
+
+- Platform metrics.
+
+  These are metrics provided by the platform that Kubernetes cluster itself runs
+  atop.
+
+  This can be things like:
+
+  - AWS/GCP/Azure/Other cloud metrics - like VPC, RDS metrics and etc.
+  - For baremetal installations - metrics from hardware loadbalancers, external
+    power/cooling solutions, hypervisors, etc.
+
+- External metrics.
```

Comment:
Right, this really isn't specific to Vector's K8s integration. An operator can use K8s as a way to deploy Vector, but it has nothing to do with k8s itself, right?

Reply:
True, this is equally valid for any kind of deployment with external metrics, nothing to do with Kubernetes. I kind of mention these for completion.